### PR TITLE
RavenDB-17674 RavenDB and Prometheus integration

### PR DESCRIPTION
**Pull request by ml054**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17674 

### Additional description

Exporter for Prometheus 

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- It has been verified by manual testing

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Web/System/MetricsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+            public static readonly string IndexType = FormatEnumHelp<IndexType>();
+            
+            private static string FormatEnumHelp<TEnum>() where TEnum : struct, Enum
+            {
+                var values = Enum.GetValues<TEnum>();
+                return "Values: " + string.Join(", ", values.Select(x => ((int)(object)x) + " = " + x));
+            }
+        }
+
+        [RavenAction("/metrics", "GET", AuthorizationStatus.Operator)]
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/MetricsHandler.cs]: `/admin/monitoring/v1/prometheus` or `/admin/monitoring/prometheus`
depending if this needs to be versioned or not

File: src/Raven.Server/Web/System/MetricsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+        public async Task Metrics()
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/MetricsHandler.cs]: this warning is an indication that you probably are not using async write to the response stream

File: src/Raven.Server/Web/System/MetricsHandler.cs [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+        {
+            return input.Replace("\"", "\\\"");
+        }
+
+        private List<DocumentDatabase> GetDatabases()
+        {
+            var databases = new List<DocumentDatabase>();
+            var landlord = ServerStore.DatabasesLandlord;
+
+            foreach (Task<DocumentDatabase> value in landlord.DatabasesCache.Values)
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/MetricsHandler.cs]: this is a bad idea, locks the cache, use regular enumerator
`foreach (var kvp in landlord.DatabasesCache)`

File: src/Raven.Server/Web/System/MetricsHandler.cs [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+                        serverMetrics.Disk.RemainingStorageSpacePercentage);
+
+                    // license 
+                    WriteGaugeWithHelp(writer, "Server license type, " + EnumHelp.LicenseType, "license_type", (int)serverMetrics.License.Type);
+                    WriteGaugeWithHelp(writer, "Server license expiration left", "license_expiration_left_seconds", serverMetrics.License.ExpirationLeftInSec);
+                    WriteGaugeWithHelp(writer, "Server license utilized CPU cores", "license_utilized_cpu_cores", serverMetrics.License.UtilizedCpuCores);
+                    WriteGaugeWithHelp(writer, "Server license max CPU cores", "license_max_cores", serverMetrics.License.MaxCores);
+                }
+
+                ms.WriteTo(ResponseBodyStream());
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/MetricsHandler.cs]: we should use async method

probably
```
ms.Position = 0;
ms.CopyToAsync(response);
```

File: src/Raven.Server/Web/System/MetricsHandler.cs [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+                    WriteGauges(writer, "Disk IO Read operations", "database_storage_io_read_operations", metrics, x => x.Storage.IoReadOperations, cachedTags);
+                    WriteGauges(writer, "Disk IO Write operations", "database_storage_io_write_operations", metrics, x => x.Storage.IoWriteOperations, cachedTags);
+                    WriteGauges(writer, "Disk Read Throughput", "database_storage_read_throughput_bytes", metrics, x => KiloBytesToBytes(x.Storage.ReadThroughputInKb),
+                        cachedTags);
+                    WriteGauges(writer, "Disk Write Throughput", "database_storage_write_throughput_bytes", metrics, x => KiloBytesToBytes(x.Storage.WriteThroughputInKb),
+                        cachedTags);
+                    WriteGauges(writer, "Disk Queue length", "database_storage_queue_length", metrics, x => x.Storage.QueueLength, cachedTags);
+                }
+
+                ms.WriteTo(ResponseBodyStream());
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/MetricsHandler.cs]: you should retrieve that response stream once, do it in handler, remember to use using

**Comment by ppekrol** [comment]
Managed to successfully setup the Prometheus and connect to this endpoint. LGTM.

Wondering if we want/need to expose numbers in this format or it should be rounded?
```
# TYPE ravendb_cpu_machine_usage gauge
ravendb_cpu_machine_usage 1.0714285714285714
```

File: src/Raven.Server/Web/System/MetricsHandler.cs (hunk 1 above)
Comment (ml054) [reply src/Raven.Server/Web/System/MetricsHandler.cs]: done - I will use versioned

File: src/Raven.Server/Web/System/MetricsHandler.cs (hunk 3 above)
Comment (ml054) [reply src/Raven.Server/Web/System/MetricsHandler.cs]: I took this from here: https://github.com/ravendb/ravendb/blob/099db8eabcd08e1cdb691e0c41b5066c9f30fea6/src/Raven.Server/Web/System/AdminMonitoringHandler.cs#L71 So I change it in both places.

File: src/Raven.Server/Web/System/MetricsHandler.cs (hunk 4 above)
Comment (ml054) [reply src/Raven.Server/Web/System/MetricsHandler.cs]: done

**Comment by ml054** [comment]
TODO: format double values. 

DONE

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+            var skipServer = GetBoolValueQueryString("skipServerMetrics", false) ?? false;
+            var skipDatabases = GetBoolValueQueryString("skipDatabasesMetrics", false) ?? false;
+            var skipIndexes = GetBoolValueQueryString("skipIndexesMetrics", false) ?? false;
+            var skipCollections = GetBoolValueQueryString("skipCollectionsMetrics", false) ?? false;
+
+            var provider = new MetricsProvider(Server);
+
+            var databases = GetDatabases();
+
+            var responseStream = ResponseBodyStream();
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: it is always good to wrap those things in using

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+            
+            if (skipServer == false)
+            {
+                await WriteServerMetrics(provider, responseStream);
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: method name should have 'Async' suffix

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs (hunk 6 above)
Comment (ml054) [reply src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: done

File: src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs (hunk 7 above)
Comment (ml054) [reply src/Raven.Server/Web/System/AdminPrometheusMonitoringHandler.cs]: done

**Review by ppekrol — APPROVED** [review]

**Comment by sergiomcalzada** [comment]
It will be great to add the endpoint to the docs :)

**Comment by ppekrol** [comment]
Hi, we have scheduled documentation work. Thanks!

**Pull request merged**